plugins/GSdx/GSOsdManager.cpp

```diff
@@ -105,59 +106,157 @@ void GSOsdManager::upload_texture_atlas(GSTexture* t) {
 
 		x += c_info[i].bw;
 	}
+
 }
 
 void GSOsdManager::Log(std::string msg) {
-	m_log.push_back((log_info){0xFF00FFFF, msg});
+	m_log.push_back((log_info){0xFF00FFFF, msg, std::chrono::system_clock::time_point()});
+}
+
+void GSOsdManager::Indicate(std::string i, bool on) {
+	m_indicator[i].on = on;
+}
+
+void GSOsdManager::RenderGlyph(GSVertexPT1* dst, glyph_info g, float x, float y, float m_sx, float m_sy, uint32 color) {
+	float x2 = x + g.bl * m_sx;
+	float y2 = -y - g.bt * m_sy;
+	float w = g.bw * m_sx;
+	float h = g.bh * m_sy;
+
+	dst->p = GSVector4(x2    , -y2    , 0.0f, 0.0f);
+	dst->t = GSVector2(g.tx       , 0.0f);
+	dst->c = color;
+	++dst;
+	dst->p = GSVector4(x2 + w, -y2    , 0.0f, 0.0f);
+	dst->t = GSVector2(g.tx + g.tw, 0.0f);
+	dst->c = color;
+	++dst;
+	dst->p = GSVector4(x2    , -y2 - h, 0.0f, 0.0f);
+	dst->t = GSVector2(g.tx       , g.ty);
+	dst->c = color;
+	++dst;
+	dst->p = GSVector4(x2 + w, -y2    , 0.0f, 0.0f);
+	dst->t = GSVector2(g.tx + g.tw, 0.0f);
+	dst->c = color;
+	++dst;
+	dst->p = GSVector4(x2    , -y2 - h, 0.0f, 0.0f);
+	dst->t = GSVector2(g.tx       , g.ty);
+	dst->c = color;
+	++dst;
+	dst->p = GSVector4(x2 + w, -y2 - h, 0.0f, 0.0f);
+	dst->t = GSVector2(g.tx + g.tw, g.ty);
+	dst->c = color;
+	++dst;
+}
+
+void GSOsdManager::RenderString(GSVertexPT1* dst, std::string msg, float x, float y, float m_sx, float m_sy, uint32 color) {
+	for(const unsigned char* p = (unsigned char*)msg.c_str(); *p; p++) {
+		RenderGlyph(dst, c_info[*p], x, y, m_sx, m_sy, color);
+
+		/* Advance the cursor to the start of the next character */
+		x += c_info[*p].ax * m_sx;
+		y += c_info[*p].ay * m_sy;
+
+		dst += 6;
+	}
 }
 
-uint32 GSOsdManager::Size() {
-	uint32 sum = 0;
-	for(unsigned entry = 0; entry < m_log.size(); ++entry) {
-		sum += m_log[entry].msg.length();
+size_t GSOsdManager::Size(float m_sx, float m_sy) {
+	size_t sum = 0;
+
+	for(const auto &pair : m_monitor) {
+		sum += pair.first.size();
+		sum += pair.second.size();
+	}
+
+	auto it = m_log.begin();
+	int add = 0;
+
+	for(; it != m_log.end();) {
+		if(1 - ((m_size+2)*(it-m_log.begin()+1)) * m_sy < -1) break;
+		if(std::chrono::system_clock::now() - it->OnScreen > std::chrono::seconds(3)) ++add;
+		sum += it->msg.size();
+		++it;
+	}
+	for(; it != m_log.end() && add > 0; --add) {
+		sum += it->msg.size();
+		++it;
+	}
+
+	for(const auto it : m_indicator) {
```

Comment:
Not C++ expert here, but can't you just write something like `sum += m_indicator.size()`